source/common/buffer/watermark_buffer.cc

```diff
@@ -147,35 +150,67 @@ void WatermarkBuffer::checkHighAndOverflowWatermarks() {
+                                                absl::optional<uint32_t> current_class,
+                                                absl::optional<uint32_t> new_class) {
   ASSERT(current_class != new_class, "Expected the current_class and new_class to be different");
 
-  if (current_class == -1 && new_class >= 0) {
+  if (!current_class.has_value() && new_class >= 0u) {
     // Start tracking
-    ASSERT(!size_class_account_sets_[new_class].contains(account));
-    size_class_account_sets_[new_class].insert(account);
-  } else if (current_class >= 0 && new_class == -1) {
+    ASSERT(!size_class_account_sets_[new_class.value()].contains(account));
+    size_class_account_sets_[new_class.value()].insert(account);
+  } else if (current_class >= 0u && !new_class.has_value()) {
     // No longer track
-    ASSERT(size_class_account_sets_[current_class].contains(account));
-    size_class_account_sets_[current_class].erase(account);
+    ASSERT(size_class_account_sets_[current_class.value()].contains(account));
+    size_class_account_sets_[current_class.value()].erase(account);
   } else {
     // Moving between buckets
-    ASSERT(size_class_account_sets_[current_class].contains(account));
-    ASSERT(!size_class_account_sets_[new_class].contains(account));
-    size_class_account_sets_[new_class].insert(
-        std::move(size_class_account_sets_[current_class].extract(account).value()));
+    ASSERT(size_class_account_sets_[current_class.value()].contains(account));
+    ASSERT(!size_class_account_sets_[new_class.value()].contains(account));
+    size_class_account_sets_[new_class.value()].insert(
+        std::move(size_class_account_sets_[current_class.value()].extract(account).value()));
   }
 }
 
 void WatermarkBufferFactory::unregisterAccount(const BufferMemoryAccountSharedPtr& account,
-                                               int current_class) {
-  if (current_class >= 0) {
-    ASSERT(size_class_account_sets_[current_class].contains(account));
-    size_class_account_sets_[current_class].erase(account);
+                                               absl::optional<uint32_t> current_class) {
+  if (current_class.has_value()) {
+    ASSERT(size_class_account_sets_[current_class.value()].contains(account));
+    size_class_account_sets_[current_class.value()].erase(account);
+  }
+}
+
+void WatermarkBufferFactory::resetAccountsGivenPressure(float pressure) {
+  ASSERT(pressure >= 0.0 && pressure <= 1.0, "Provided pressure is out of range [0, 1].");
+
+  // Compute buckets to clear
+  const uint32_t buckets_to_clear = std::min<uint32_t>(
+      std::floor(pressure * BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_) + 1, 8);
+  uint32_t bucket_idx = BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_ - buckets_to_clear;
+
+  ENVOY_LOG_MISC(warn, "resetting streams in buckets >= {}", bucket_idx);
+
+  // Clear buckets
+  while (bucket_idx < BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_) {
```

Comment:
As discussed can you either limit these to an arbitrary hard-coded number of kills per dispatcher loop (say 50?) or add a TODO to do this in a follow-up?

Reply:
Follow up on this: https://github.com/KBaichoo/envoy/pull/128 -- will open PR against upstream when dependent PR submitted.